crates/core/src/factories/core_kitsune.rs

```diff
@@ -0,0 +1,109 @@
+//! The core kitsune implementation provided by Kitsune2.
+
+use kitsune2_api::{config::*, kitsune::*, *};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+const MOD_NAME: &str = "CoreKitsune";
+
+/// Configuration parameters for [CoreKitsuneFactory].
+#[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoreKitsuneConfig {}
+
+impl ModConfig for CoreKitsuneConfig {}
+
+/// The core kitsune implementation provided by Kitsune2.
+/// You probably will have no reason to use something other than this.
+/// This abstraction is mainly here for testing purposes.
+#[derive(Debug)]
+pub struct CoreKitsuneFactory {}
+
+impl CoreKitsuneFactory {
+    /// Construct a new CoreKitsuneFactory.
+    pub fn create() -> DynKitsuneFactory {
+        let out: DynKitsuneFactory = Arc::new(CoreKitsuneFactory {});
+        out
+    }
+}
+
+impl KitsuneFactory for CoreKitsuneFactory {
+    fn default_config(&self, config: &mut Config) -> K2Result<()> {
+        config
+            .add_default_module_config::<CoreKitsuneConfig>(MOD_NAME.into())?;
+        Ok(())
+    }
+
+    fn create(
+        &self,
+        builder: Arc<builder::Builder>,
+        handler: DynKitsuneHandler,
+    ) -> BoxFut<'static, K2Result<DynKitsune>> {
+        Box::pin(async move {
+            let config = builder
+                .config
+                .get_module_config::<CoreKitsuneConfig>(MOD_NAME)?;
+            let out: DynKitsune =
+                Arc::new(CoreKitsune::new(builder.clone(), config, handler));
+            Ok(out)
+        })
+    }
+}
+
+type SpaceFut =
+    futures::future::Shared<BoxFut<'static, K2Result<space::DynSpace>>>;
+type Map = HashMap<SpaceId, SpaceFut>;
+
+#[derive(Debug)]
+struct CoreKitsune {
+    builder: Arc<builder::Builder>,
+    handler: DynKitsuneHandler,
+    map: std::sync::Mutex<Map>,
+}
+
+impl CoreKitsune {
+    pub fn new(
+        builder: Arc<builder::Builder>,
+        _config: CoreKitsuneConfig,
+        handler: DynKitsuneHandler,
+    ) -> Self {
+        Self {
+            builder,
+            handler,
+            map: std::sync::Mutex::new(HashMap::new()),
+        }
+    }
+}
+
+impl Kitsune for CoreKitsune {
+    fn space(&self, space: SpaceId) -> BoxFut<'_, K2Result<space::DynSpace>> {
+        Box::pin(async move {
+            use std::collections::hash_map::Entry;
+
+            // This is quick, we don't hold the lock very long,
+            // because we're just constructing the future here,
+            // not awaiting it.
+            let fut = match self.map.lock().unwrap().entry(space.clone()) {
+                Entry::Occupied(e) => e.get().clone(),
+                Entry::Vacant(e) => {
+                    let builder = self.builder.clone();
+                    let handler = self.handler.clone();
+                    e.insert(futures::future::FutureExt::shared(Box::pin(
```

Comment:
Nothing else done with this future/map yet, right?

Reply:
Not sure anything else ever needs to be done with this map. Really space creation shouldn't ever error, and if it does it probably means the whole tokio worker thread died or something. It's just easier to propagate the errors. The shared future is just about race condition management of parallel requests for the same space that doesn't yet exist.

Reply:
And the shared future is not yet awaited anywhere, is it?

Reply:
Yep, it's awaited right here to get the result for this function call : )

Reply:
Ah, now I get the comment above. The future is constructed while the map is locked, but it's awaited afterwards when the map lock is released, and awaiting it constructs the space, and the future is shared in case some other thread created the same space in the meantime.